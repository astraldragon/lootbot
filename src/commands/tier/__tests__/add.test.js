jest.mock("discord-akairo");
jest.mock("../../../models", () => {
  return {
    Tier: { findOrCreate: jest.fn("findOrCreate") }
  };
});

const { Command } = require("discord-akairo");
const { Tier } = require("../../../models");
const { checkManagePermissions } = require("../../../support");
const AddCommand = require("../add");

describe("commands/tier/add", () => {
  beforeEach(() => {
    this.command = new AddCommand();

    this.args = {
      name: "Tier 1",
      color: "#ffffff",
      image: { href: "http://someimage.png" },
      weight: 5,
      luckyWeight: 10
    };

    this.msg = {
      guild: { id: "1234" },
      channel: {
        send: jest.fn("send").mockReturnValue(Promise.resolve())
      }
    };
  });

  it("should configure properly", () => {
    expect(Command).toMatchSnapshot();
  });

  it("should have the proper permission check", () => {
    expect(Command.mock.calls[0][1].options.permissions).toBe(
      checkManagePermissions
    );
  });

  it("should create the tier", async () => {
    let { image, ...defaults } = this.args;

    Tier.findOrCreate.mockReturnValue([{}, true]);
    await this.command.exec(this.msg, this.args);

    expect(Tier.findOrCreate).toHaveBeenCalledWith({
      where: { name: this.args.name, guild: this.msg.guild.id },
      defaults: { ...defaults, image: image.href, guild: this.msg.guild.id }
    });

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} added`
    );
  });

  it("should notify when the loot added already exists", async () => {
    Tier.findOrCreate.mockReturnValue([{}, false]);
    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} already exists`
    );
  });

  it("should notify if any errors occur", async () => {
    Tier.findOrCreate.mockImplementation(() => {
      throw new Error("test");
    });

    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `An error occurred adding ${this.args.name}`
    );
  });
});
