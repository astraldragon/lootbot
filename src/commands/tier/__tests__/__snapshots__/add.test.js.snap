// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/tier/add should configure properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "tier-add",
      Object {
        "aliases": Array [
          "tier-add",
          "ta",
        ],
        "args": Array [
          Object {
            "id": "name",
            "prompt": Object {
              "start": "What is the name of the tier?",
            },
            "type": "string",
          },
          Object {
            "id": "color",
            "prompt": Object {
              "start": "What is the color of the tier?",
            },
            "type": "color",
          },
          Object {
            "id": "image",
            "prompt": Object {
              "start": "What is the image of the tier?",
            },
            "type": "url",
          },
          Object {
            "id": "weight",
            "prompt": Object {
              "start": "What is the weight of the loot?",
            },
            "type": "number",
          },
          Object {
            "id": "luckyWeight",
            "prompt": Object {
              "start": "What is the lucky weight of the loot?",
            },
            "type": "number",
          },
        ],
        "category": "Tier",
        "channelRestriction": "guild",
        "description": Object {
          "content": "Add tier to the lootbox",
          "examples": Array [
            "tier:add Common #123456 https://some.pathtoimage.png 25 75",
          ],
          "usage": "<name> <color> <image> <weight> <luckyWeight>",
        },
        "options": Object {
          "permissions": [Function],
        },
        "split": "quoted",
      },
    ],
  ],
}
`;
